src/common/commonutils/PerfUtils.c

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "Internal.h"
+
+int StartPerfClock(PERF_CLOCK* clock, void* log)
+{
+    int status = EINVAL;
+
+    if (NULL == clock)
+    {
+        OsConfigLogError(log, "StartPerfClock called with an clock invalid argument");
+        return status;
+    }
+
+    memset(clock, 0, sizeof(PERF_CLOCK));
+
+    if (0 != (status = clock_gettime(CLOCK_MONOTONIC, &(clock->start))))
+    {
+        OsConfigLogError(log, "StartPerfClock: clock_gettime failed with %d (%d)", status, errno);
+    }
+
+    return status;
+}
+
+int StopPerfClock(PERF_CLOCK* clock, void* log)
+{
+    int status = EINVAL;
+
+    if (NULL == clock)
+    {
+        OsConfigLogError(log, "StopPerfClock called with an invalid clock argument");
+        return status;
+    }
+
+    if (0 == (status = clock_gettime(CLOCK_MONOTONIC, &(clock->stop))))
+    {
+        if (clock->stop.tv_sec < clock->start.tv_sec)
+        {
+            OsConfigLogError(log, "StopPerfClock: clock_gettime returned an earlier time than expected (%ld seconds earlier)", 
+                clock->start.tv_sec - clock->stop.tv_sec);
+
+            memset(clock, 0, sizeof(PERF_CLOCK));
+
+            status = ENOENT;
+        }
+    }
+    else
+    {
+        OsConfigLogError(log, "StopPerfClock: clock_gettime failed with %d (%d)", status, errno);
+    }
+
+    return status;
+}
+
+long GetPerfClockTime(PERF_CLOCK* clock, void* log)
+{
+    long microseconds = -1;
+
+    if ((NULL == clock) || (0 == clock->stop.tv_sec))
+    {
+        OsConfigLogError(log, "GetPerfClockTime called with an invalid clock argument");
+        return microseconds;
+    }
+
+    microseconds = ((clock->stop.tv_sec - clock->start.tv_sec) * 1000000) + 
+        (((clock->stop.tv_nsec > clock->start.tv_nsec) ? (clock->stop.tv_nsec - clock->start.tv_nsec) : 0) / 1000);
```

Comment:
This calculation does not give the right answer. Take the following example:
```
start: tv_sec=54217 tv_nsec=999942868
stop: tv_sec=54218 tv_nsec=18649
```

The correct answer is `76us` but the computation results in `1000000us`.

Reply:
Excellent find, see please last iteration for a fix.